encoding/thrift/outbound.go

```diff
@@ -0,0 +1,148 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package thrift
+
+import (
+	"bytes"
+	"io/ioutil"
+
+	"github.com/yarpc/yarpc-go"
+	"github.com/yarpc/yarpc-go/transport"
+
+	"github.com/uber/thriftrw-go/protocol"
+	"github.com/uber/thriftrw-go/wire"
+	"golang.org/x/net/context"
+)
+
+// Client is a generic Thrift client. It speaks in raw Thrift payloads. The code
+// generator is responsible for putting a pretty interface in front of it.
+type Client interface {
+	Call(ctx context.Context, methon string, r *Request) (*Response, yarpc.Meta, error)
+}
+
+// Config contains the configuration for the Client.
+type Config struct {
+	// Name of the Thrift service.
+	Service string
+
+	// Outbound through which requests will be sent. Required.
+	Outbound transport.Outbound
+
+	// Thrift encoding protocol. Defaults to Binary if nil.
+	Protocol protocol.Protocol
+}
+
+// New creates a new Thrift client.
+func New(c Config) Client {
+	// Code generated for Thrift client instantiation will probably be something
+	// like this:
+	//
+	// 	func New(t transport.Outbound) *MyServiceClient {
+	// 		c := thrift.New(thrift.Config{
+	// 			Service: "MyService",
+	// 			Outbound: t,
+	// 			Protocol: protocol.Binary,
+	// 		})
+	// 		return &MyServiceClient{client: c}
+	// 	}
+	//
+	// So Config is really the internal config as far as consumers of the
+	// generated client are concerned.
+
+	p := c.Protocol
+	if p == nil {
+		p = protocol.Binary
+	}
+
+	return thriftClient{
+		p:       p,
+		t:       c.Outbound,
+		service: c.Service,
+	}
+}
+
+type thriftClient struct {
+	t transport.Outbound
+	p protocol.Protocol
+
+	service string
+}
+
+func (t thriftClient) Call(ctx context.Context, method string, r *Request) (*Response, yarpc.Meta, error) {
+	// Code generated for Thrift client calls will probable be something like
+	// this:
+	//
+	// 	func (c *MyServiceClient) someMethod(ctx context.Context, m yarpc.Meta, arg1 Arg1Type, arg2Type) (returnValue, yarpc.Meta, error) {
+	// 		args := someMethodArgs{arg1: arg1, arg2: arg2}
+	// 		resp, m, err := c.client.Call(ctx, "someMethod", &thrift.Request{
+	// 			Meta: m,
+	// 			Body: args.ToWire(),
+	// 		})
+	// 		if err != nil { return nil, m, err }
+	// 		if resp.Exception1 != nil {
+	// 			return nil, m, resp.Exception1
```

Comment:
If we don't use `IsGetValueError` but the `GetValueError` wrapper, we should be able to handle unexpected errors but it will require users to *always* get the underlying error from the `GetValueError` rather than type matching on it directly. Alternatively, `IsGetValueError` could always return true when the error matches `thriftrw.UnexpectedError` with the assumption that it's from the future.